app/models/ems_cluster.rb

```diff
@@ -182,26 +182,11 @@ def parent_datacenter
     detect_ancestor(:of_type => 'EmsFolder') { |a| a.kind_of?(Datacenter) }
   end
 
-  def event_where_clause(assoc = :ems_events)
-    return ["ems_cluster_id = ?", id] if assoc.to_sym == :policy_events
-
-    cond = ["ems_cluster_id = ?"]
-    cond_params = [id]
-
-    ids = host_ids
-    unless ids.empty?
-      cond << "host_id IN (?) OR dest_host_id IN (?)"
-      cond_params += [ids, ids]
-    end
-
-    ids = vm_or_template_ids
-    unless ids.empty?
-      cond << "vm_or_template_id IN (?) OR dest_vm_or_template_id IN (?)"
-      cond_params += [ids, ids]
-    end
-
-    cond_params.unshift(cond.join(" OR ")) unless cond.empty?
-    cond_params
+  def event_where_clause_ems_events
+    super
+      .or(EmsEvent.where(:host_id => host_ids))
```

Comment:
```suggestion
      .or(EmsEvent.where(:host_id => host_ids))
      .or(EmsEvent.where(:dest_host_id => host_ids))
```

Reply:
these OR clauses force us to do a table scan. and for such a large table... darn

Reply:
I don't know why we have all of these columns on this table.

Reply:
Events can link back to all kinds of different objects.  Your first inclination might be to use polymorphic, but an even might be related to a vm, host, and cluster simultaneously.  So, we added foreign keys for each.  In addition for clone and copy operations you can have a source and dest host or source and dest vm.